From 21109ed6c59bd15449ca15c2fd2d99f2c8b135c8 Mon Sep 17 00:00:00 2001
Message-Id: <21109ed6c59bd15449ca15c2fd2d99f2c8b135c8.1449508440.git.jen@redhat.com>
From: Alex Williamson <alex.williamson@redhat.com>
Date: Thu, 3 Dec 2015 22:49:10 -0600
Subject: [CHANGE] Include port in HTTP Host header as needed
To: rhvirt-patches@redhat.com,
    jen@redhat.com

RH-Author: Alex Williamson <alex.williamson@redhat.com>
Message-id: <20151203223945.7811.19065.stgit@gimli.home>
Patchwork-id: 68504
O-Subject: [RHEL6.8 gpxe PATCH] [http] Include port in HTTP Host header as needed
Bugzilla: 1231931
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Gerd Hoffmann <kraxel@redhat.com>
RH-Acked-by: Jeff Nelson <jenelson@redhat.com>

From: Malte Starostik <lists@malte.homeip.net>

Bugzilla: 1231931
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=10185370
Upstream: 69b7d57265679d76e26581d034e8f8ab5168bb27 (ipxe.git)

According to section 14.23 of RFC2616, an HTTP Host header without
port implies the default port is used.  Thus, when fetching from
anywhere but port 80 for HTTP or 443 for HTTPS, the port ought to be
explicitly given in that header.  Otherwise, some servers might fail
to associate the request with the correct virtual host or generate
incorrect self-referencing URLs.

Signed-off-by: Michael Brown <mcb30@ipxe.org>
Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
Signed-off-by: Jeff E. Nelson <jen@redhat.com>
---
 src/net/tcp/http.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/net/tcp/http.c b/src/net/tcp/http.c
index 93ccfd3..6a33eb1 100644
--- a/src/net/tcp/http.c
+++ b/src/net/tcp/http.c
@@ -430,7 +430,7 @@ static void http_step ( struct process *process ) {
 					  "GET %s%s%s HTTP/1.0\r\n"
 					  "User-Agent: gPXE/" VERSION "\r\n"
 					  "%s%s%s"
-					  "Host: %s\r\n"
+					  "Host: %s%s%s\r\n"
 					  "\r\n",
 					  ( path ? path : "/" ),
 					  ( query ? "?" : "" ),
@@ -439,7 +439,12 @@ static void http_step ( struct process *process ) {
 					    "Authorization: Basic " : "" ),
 					  ( user ? user_pw_base64 : "" ),
 					  ( user ? "\r\n" : "" ),
-					  host ) ) != 0 ) {
+					  host,
+					  ( http->uri->port ?
+					    ":" : "" ),
+					  ( http->uri->port ?
+					    http->uri->port : "" )
+					  ) ) != 0 ) {
 			http_done ( http, rc );
 		}
 	}
-- 
2.5.0

